internal/handler/msg_find.go

```diff
@@ -71,20 +69,20 @@ func (h *Handler) MsgFind(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, er
 	}
 
 	var cList *backends.ListCollectionsResult
-
-	if cList, err = db.ListCollections(ctx, nil); err != nil {
+	collectionParam := backends.ListCollectionsParams{Name: params.Collection}
+	if cList, err = db.ListCollections(ctx, &collectionParam); err != nil {
 		return nil, err
 	}
 
 	var cInfo backends.CollectionInfo
 
-	// TODO https://github.com/FerretDB/FerretDB/issues/3601
-	if i, found := slices.BinarySearchFunc(cList.Collections, params.Collection, func(e backends.CollectionInfo, t string) int {
-		return cmp.Compare(e.Name, t)
-	}); found {
-		cInfo = cList.Collections[i]
+	if len(cList.Collections) == 0 {
+		msg := fmt.Sprintf("Invalid collection name: %s", collectionParam.Name)
+		return nil, handlererrors.NewCommandErrorMsgWithArgument(handlererrors.ErrInvalidNamespace, msg, document.Command())
```

Comment:
I think this change broken a lot of integration tests

Reply:
found a logical issue in error handling. fixed the same in all handlers.